delfin/drivers/hpe/hpe_3par/component_handler.py

```diff
@@ -211,3 +214,198 @@ def list_volumes(self, context):
+                serial_number = None
+                manufacturer = None
+                model = None
+                firmware = None
+                if disks_inventory_map:
+                    inventory_map = disks_inventory_map.get(disk_id)
+                    if inventory_map:
+                        serial_number = inventory_map.get('disk_serial')
+                        manufacturer = inventory_map.get('disk_mfr')
+                        model = inventory_map.get('disk_model')
+                        firmware = inventory_map.get('disk_fw_rev')
+                speed = None
+                if disk.get('rpm'):
+                    speed = int(disk.get('rpm')) * units.k
+                disk_model = {
+                    'name': disk.get('cagepos'),
+                    'storage_id': storage_id,
+                    'native_disk_id': disk_id,
+                    'serial_number': serial_number,
+                    'manufacturer': manufacturer,
+                    'model': model,
+                    'firmware': firmware,
+                    'speed': speed,
+                    'capacity': capacity,
+                    'status': status,
+                    'physical_type': consts.DISK_PHYSICAL_TYPE_MAP.get(
+                        disk.get('type').upper(),
+                        constants.DiskPhysicalType.UNKNOWN),
+                    'logical_type': None,
+                    'health_score': None,
+                    'native_disk_group_id': None,
+                    'location': disk.get('cagepos')
+                }
+                disk_list.append(disk_model)
+        return disk_list
+
+    def list_ports(self, storage_id):
+        ports = self.ssh_handler.get_ports()
+        port_list = []
+        if ports:
+            ports_inventory_map = self.ssh_handler.get_ports_inventory()
+            ports_config_map = self.ssh_handler.get_ports_config()
+            ports_iscsi_map = self.ssh_handler.get_ports_iscsi()
+            ports_rcip_map = self.ssh_handler.get_ports_rcip()
+            ports_connected_map = self.ssh_handler.get_ports_connected()
+            ports_fcoe_map = self.ssh_handler.get_ports_fcoe()
+            port_fs_map = self.ssh_handler.get_ports_fs()
+            for port in ports:
+                port_id = port.get('n:s:p')
+                port_type = ''
+                if ports_inventory_map:
+                    port_type = ports_inventory_map.get(port_id, '')
+                max_speed = ''
+                if ports_config_map:
+                    max_speed = ports_config_map.get(port_id, '')
+                ip_addr = None
+                ip_mask = None
+                ipv4 = None
+                ipv4_mask = None
+                ipv6 = None
+                ipv6_mask = None
+                rate = ''
+                if ports_connected_map:
+                    rate = ports_connected_map.get(port_id, '')
+                if not ip_addr and ports_iscsi_map:
```

Comment:
"not ip_addr" can be placed outside

Reply:
Each map extracts different values, and when a value is found in one of the maps, there is no need to extract values from other maps. If you add NOT IP_ADDR to the outermost layer, it will look up each map once.